### PR TITLE
[feat] Adding Custom Authenticator and Http Conn Pool

### PR DESCRIPTION
**Pull request by andrefurlan-db**
This PR adds:
* the ability to pass in an Authenticator to the driver. If a token is passed, the PAT authenticator will be used
* reusability of HTTP connections, including increasing MaxIdleConnsPerHost to 10.
* better error message, including Databricks specific error headers.
* fixed a few bugs in the way, namely on the retryable client and failure to close operation when context is canceled.

### CODE REVIEW

File: auth/auth.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package auth
+
+import "net/http"
+
+type Authenticator interface {
```
Comment (moderakh) [inline auth/auth.go:5]: is this a public user facing API?

Comment (andrefurlan-db) [reply auth/auth.go:5]: It is a public facing interface. Only needed if user wants to have their own implementation.

File: connection.go [hunk 2]
```diff
@@ -102,7 +102,7 @@ func (c *conn) ExecContext(ctx context.Context, query string, args []driver.Name
 		// since we have an operation handle we can close the operation if necessary
 		alreadyClosed := exStmtResp.DirectResults != nil && exStmtResp.DirectResults.CloseOperation != nil
 		if !alreadyClosed && (opStatusResp == nil || opStatusResp.GetOperationState() != cli_service.TOperationState_CLOSED_STATE) {
-			_, err1 := c.client.CloseOperation(ctx, &cli_service.TCloseOperationReq{
+			_, err1 := c.client.CloseOperation(context.Background(), &cli_service.TCloseOperationReq{
```
Comment (rcypher-databricks) [inline connection.go:105]: nice catch

**Review by rcypher-databricks — APPROVED** [review]

**Comment by andrefurlan-db** [comment]
This is a v1.1.0 release PR

File: auth/auth.go (hunk 1 above)
Comment (moderakh) [reply auth/auth.go:5]: @andrefurlan-db I am planning to write the decision doc we discussed sometime end of this week / early next week. as this is a change in the public surface area, can we wait to close on that decision before merging this?

Comment (andrefurlan-db) [reply auth/auth.go:5]: This PR is target for version 1.1.0. It won't be merged this year.

File: auth/pat/pat.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package pat
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/pkg/errors"
+)
+
+type PATAuth struct {
```
Comment (nfx) [inline auth/pat/pat.go:10]: why do you need to re-implement auth, if there's a concurrent [PR](https://github.com/databricks/databricks-sql-go/pull/63) on integrating with Go SDK, that solves the auth problem in a unified way across different tools?

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by moderakh — CHANGES_REQUESTED** [review]
pending a one decision doc on the plan as a follow up internal discussion.

**Pull request closed without merging**